### PR TITLE
CoreFX FileSystemWatcher (Linux)

### PR DESCRIPTION
**Pull request by alexischr**
Progresses #6966

System.Native icalls now live in `metadata/pal-icalls.c|h`

### CODE REVIEW

File: mcs/class/System/linux_net_4_x_System.dll.sources [hunk 1]
```diff
@@ -12,6 +12,6 @@
 ../../../external/corefx/src/Common/src/Interop/Unix/System.Native/Interop.Stat.cs
 ../../../external/corefx/src/Common/src/Interop/Unix/System.Native/Interop.Poll.cs
 ../../../external/corefx/src/Common/src/Interop/Unix/System.Native/Interop.Open.cs
-../../../external/corefx/src/Common/src/Interop/Unix/System.Native/Interop.Read.cs
+corefx/Interop.Read.cs
```
Comment (marek-safar) [inline mcs/class/System/linux_net_4_x_System.dll.sources]: We should keep the Unix prefix

File: mcs/class/System/corefx/Interop.Read.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                for (int i = 0; i < bytes_read; i++)
+                    buffer[i] = inbuf[i];
+            }
+
+            Marshal.FreeHGlobal (unmanaged_buffer);
+            return bytes_read;
+        }
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        internal static extern unsafe int Read_Internal(SafeFileHandle fd, IntPtr buffer, int count);
```
Comment (marek-safar) [inline mcs/class/System/corefx/Interop.Read.cs]: Why internal?

File: mcs/class/System/corefx/Interop.Read.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+using System;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
```
Comment (marek-safar) [inline mcs/class/System/corefx/Interop.Read.cs]: Better to add a comment where this was copied from (if it was)

File: mono/metadata/pal-icalls.c [hunk 4]
```diff
@@ -0,0 +1,102 @@
+/**
+ * \file
+ * System.Native PAL internal calls
```
Comment (marek-safar) [inline mono/metadata/pal-icalls.c]: This needs better description what this file is about

File: mono/metadata/pal-icalls.c [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+		mono_thread_info_uninstall_interrupt (&interrupted);
+	} while (!interrupted && result < 0 && errno == EINTR);
+
+	if (interrupted)
+		return -1;
+
+	return result;
+}
+
+#if defined(__APPLE__)
```
Comment (marek-safar) [inline mono/metadata/pal-icalls.c]: @luhenry @kumpera do we want to mix unix and apple code in the same file?

File: mono/metadata/pal-icalls.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+#ifndef __MONO_METADATA_ICALL_PAL_H__
+#define __MONO_METADATA_ICALL_PAL_H__
+
+#include <metadata/object-internals.h>
+
+#if !defined(HOST_WIN32) && !defined(HOST_WASM) /* Unix System.Native PAL */
+gint32 ves_icall_CoreFX_Interop_Sys_Read (MonoSafeHandleHandle fd_handle, gchar* buffer, gint32 count);
+
+#if defined(__APPLE__)
```
Comment (marek-safar) [inline mono/metadata/pal-icalls.h]: I'd move this out of the Unix like define

File: mcs/class/System/corefx/Interop.Read.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        /// Note - on fail. the position of the stream may change depending on the platform; consult man 2 read for more info
+        /// </returns>
+        internal static unsafe int Read (SafeFileHandle fd, byte* buffer, int count)
+        {
+            int bytes_read = -1;
+            IntPtr unmanaged_buffer = Marshal.AllocHGlobal (count);
+            bool has_ref = false;
+            try {
+                fd.DangerousAddRef (ref has_ref);
+                if (has_ref)
```
Comment (luhenry) [inline mcs/class/System/corefx/Interop.Read.cs]: `has_ref` will always be true if `fd.DangerousAddRef` succeeded so no need to check for it here. And the convention is to name it `release`.

File: mcs/class/System/corefx/Interop.Read.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        /// <param name="buffer">The buffer to read info into</param>
+        /// <param name="count">The size of the buffer</param>
+        /// <returns>
+        /// Returns the number of bytes read on success; otherwise, -1 is returned
+        /// Note - on fail. the position of the stream may change depending on the platform; consult man 2 read for more info
+        /// </returns>
+        internal static unsafe int Read (SafeFileHandle fd, byte* buffer, int count)
+        {
+            int bytes_read = -1;
+            IntPtr unmanaged_buffer = Marshal.AllocHGlobal (count);
```
Comment (luhenry) [inline mcs/class/System/corefx/Interop.Read.cs]: Why are you allocating a temporary buffer and not using `buffer`? There is no need for it since the expectation is: if the returned value is <= 0, then any data put into `buffer` is invalid.

File: mono/dis/Makefile.am [hunk 9]
```diff
@@ -7,6 +7,7 @@ export HOST_CC
 endif
 
 glib_libs = $(top_builddir)/mono/eglib/libeglib.la
+system_native_lib = $(top_builddir)/mono/metadata/libmono-system-native.la
```
Comment (luhenry) [inline mono/dis/Makefile.am]: `dis` shouldn't rely on `System.Native`.

File: mono/metadata/icall-def.h [hunk 10]
```diff
@@ -65,8 +65,13 @@
  */
 
  #if defined(__APPLE__)
-ICALL_TYPE(CLR_INTEROP, "Interop/RunLoop", CLR_INTEROP_1)
-ICALL(CLR_INTEROP_1, "CFRunLoopRun", ves_icall_CoreFX_Interop_RunLoop_CFRunLoopRun)
+ICALL_TYPE(INTEROP_RUNLOOP, "Interop/RunLoop", INTEROP_RUNLOOP_1)
+ICALL(INTEROP_RUNLOOP_1, "CFRunLoopRun", ves_icall_CoreFX_Interop_RunLoop_CFRunLoopRun)
+#endif
+
+#if !defined(HOST_WIN32) && !defined(HOST_WASM)
```
Comment (luhenry) [inline mono/metadata/icall-def.h]: ` && !defined(ENABLE_MONODROID) && !defined(ENABLE_MONOTOUCH)`

File: mono/metadata/pal-icalls.c [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+#include <config.h>
+#include <glib.h>
+#include <mono/metadata/handle.h>
+#include <mono/utils/hazard-pointer.h>
+#include <mono/utils/mono-threads.h>
+
+#include "pal-icalls.h"
+
+#if !defined(HOST_WIN32) && !defined(HOST_WASM) /* Unix System.Native PAL */
```
Comment (luhenry) [inline mono/metadata/pal-icalls.c]: ` && !defined(ENABLE_MONODROID) && !defined(ENABLE_MONOTOUCH)`

File: mono/metadata/pal-icalls.c [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+extern int32_t SystemNative_Read(intptr_t fd, void* buffer, int32_t bufferSize);
```
Comment (luhenry) [inline mono/metadata/pal-icalls.c]: This needs to be in `pal-icalls.h`.

File: mono/metadata/pal-icalls.c [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+	mono_threads_suspend_abort_syscall (info);
+	mono_threads_wait_pending_operations ();
+
+	mono_hazard_pointer_clear (hp, 1);
+	mono_threads_end_global_suspend ();
+	mono_thread_info_suspend_unlock ();
+}
+
+gint32
+ves_icall_CoreFX_Interop_Sys_Read (MonoSafeHandleHandle fd_handle, gchar* buffer, gint32 count)
```
Comment (luhenry) [inline mono/metadata/pal-icalls.c]: `ves_icall_Interop_Sys_Read`.

File: mono/metadata/pal-icalls.h [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+*/
+
+#ifndef __MONO_METADATA_ICALL_PAL_H__
+#define __MONO_METADATA_ICALL_PAL_H__
+
+#include <metadata/object-internals.h>
+
+#if !defined(HOST_WIN32) && !defined(HOST_WASM) /* Unix System.Native PAL */
+gint32 ves_icall_CoreFX_Interop_Sys_Read (MonoSafeHandleHandle fd_handle, gchar* buffer, gint32 count);
```
Comment (luhenry) [inline mono/metadata/pal-icalls.h]: `ves_icall_Interop_Sys_Read`.

File: mono/metadata/pal-icalls.h [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+#if defined(__APPLE__)
+void ves_icall_CoreFX_Interop_RunLoop_CFRunLoopRun (void);
```
Comment (luhenry) [inline mono/metadata/pal-icalls.h]: `ves_icall_Interop_RunLoop_CFRunLoopRun`.

File: mono/metadata/pal-icalls.h [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+ * Copyright 2018 Microsoft
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+*/
+
+#ifndef __MONO_METADATA_ICALL_PAL_H__
+#define __MONO_METADATA_ICALL_PAL_H__
+
+#include <metadata/object-internals.h>
+
+#if !defined(HOST_WIN32) && !defined(HOST_WASM) /* Unix System.Native PAL */
```
Comment (luhenry) [inline mono/metadata/pal-icalls.h]: ` && !defined(ENABLE_MONODROID) && !defined(ENABLE_MONOTOUCH)`

File: mono/metadata/pal_config.h [hunk 17]
```diff
@@ -1 +1,2 @@
+#define MONO 1
```
Comment (luhenry) [inline mono/metadata/pal_config.h]: Put it after the `#include` and add an empty line in between.

File: mono/mini/Makefile.am.in [hunk 18]
```diff
@@ -26,13 +26,15 @@ boehm_libs=	\
 	$(monodir)/mono/metadata/libmonoruntime.la	\
 	$(monodir)/mono/utils/libmonoutils.la \
 	$(glib_libs) \
-	$(libgc_libs)
+	$(libgc_libs)	\
+	$(monodir)/mono/metadata/libmono-system-native.la
```
Comment (luhenry) [inline mono/mini/Makefile.am.in]: `mono-boehm` shouldn't rely on `System.Native`.

File: mono/mini/Makefile.am.in [hunk 19]
```diff
@@ -26,13 +26,15 @@ boehm_libs=	\
...
 
 sgen_libs = \
 	$(monodir)/mono/metadata/libmonoruntimesgen.la	\
 	$(monodir)/mono/sgen/libmonosgen.la	\
 	$(monodir)/mono/utils/libmonoutils.la \
-	$(glib_libs)
+	$(glib_libs)	\
+	$(monodir)/mono/metadata/libmono-system-native.la
```
Comment (luhenry) [inline mono/mini/Makefile.am.in]: `mono-sgen` shouldn't rely on `System.Native`.

File: tools/monograph/Makefile.am [hunk 20]
```diff
@@ -36,7 +36,8 @@ AM_CPPFLAGS = 				\
 monograph_LDADD = \
 	$(runtime_lib)			\
 	$(glib_libs)			\
-	$(LLVM_LIBS)
+	$(LLVM_LIBS)	\
+	$(top_builddir)/mono/metadata/libmono-system-native.la
```
Comment (luhenry) [inline tools/monograph/Makefile.am]: `monograph` shouldn't rely on `System.Native`.

File: tools/pedump/Makefile.am [hunk 21]
```diff
@@ -24,7 +24,8 @@ pedump_LDADD = 			\
 	$(top_builddir)/mono/utils/libmonoutils.la \
 	$(LLVM_LIBS)			\
 	$(LLVM_LDFLAGS)			\
-	$(glib_libs)
+	$(glib_libs)	\
+	$(top_builddir)/mono/metadata/libmono-system-native.la
```
Comment (luhenry) [inline tools/pedump/Makefile.am]: `pedump` shouldn't rely on `System.Native`.

File: mono/metadata/pal-icalls.c [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+
+static void
+interrupt_CFRunLoop (gpointer data)
+{
+	g_assert (data);
+	CFRunLoopStop (data);
+}
+
+void
+ves_icall_CoreFX_Interop_RunLoop_CFRunLoopRun (void)
```
Comment (luhenry) [inline mono/metadata/pal-icalls.c]: `ves_icall_Interop_RunLoop_CFRunLoopRun`.

File: mono/metadata/pal-icalls.c [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+
+gint32
+ves_icall_CoreFX_Interop_Sys_Read (MonoSafeHandleHandle fd_handle, gchar* buffer, gint32 count)
+{
+	gint32 result;
+	gpointer fd = MONO_HANDLE_GETVAL(fd_handle, handle);
+	gboolean interrupted;
+
+	do {
+		mono_thread_info_install_interrupt (interrupt_read, (gpointer) (gsize) mono_native_thread_id_get (), &interrupted);
```
Comment (luhenry) [inline mono/metadata/pal-icalls.c]: This `install`+`uninstall` is not needed since a signal will interrupt the syscall and that's what you are trying to do in `interrupt_read`. Why not doing the `while` loop in managed instead?

File: mono/metadata/pal-icalls.h (hunk 6 above)
Comment (luhenry) [reply mono/metadata/pal-icalls.h]: No.

File: mono/metadata/pal-icalls.c (hunk 5 above)
Comment (luhenry) [reply mono/metadata/pal-icalls.c]: That's fine as long as the code is not overblown with preprocessor conditionals. In this case, it's definitely manageable.

File: mcs/class/System/linux_net_4_x_System.dll.sources (hunk 1 above)
Comment (luhenry) [reply mcs/class/System/linux_net_4_x_System.dll.sources]: Why? The original file is named `Interop.Read.cs` and the existing convention is to take the name of the class or file we are copying.

Comment (marek-safar) [reply mcs/class/System/linux_net_4_x_System.dll.sources]: Because the original file is for Unix only as it's located at `/Unix/System.Native/Interop.Read.cs` we don't have that convention yet as this is our first platform specific PAL file we are customizing.

Comment (luhenry) [reply mcs/class/System/linux_net_4_x_System.dll.sources]: Ok.

File: mono/metadata/pal-icalls.c (hunk 23 above)
Comment (alexischr) [reply mono/metadata/pal-icalls.c]: @luhenry I don't understand this point. We want the runtime itself to be able to interrupt the read, so that it can shut down a thread blocked in `read`. If the runtime doesn't know to do this, it will not be interrupted and the process will hang on exit. The purpose of using an icall for this was to allow this to happen.

File: mcs/class/System/corefx/Interop.Read.cs (hunk 2 above)
Comment (alexischr) [reply mcs/class/System/corefx/Interop.Read.cs]: Just following a BCL "convention" where if part of an implementation relies on an icall then it ends with "_Internal". I don't mind changing it

File: mono/metadata/pal-icalls.h (hunk 6 above)
Comment (alexischr) [reply mono/metadata/pal-icalls.h]: I went back and forth on this and decided to keep it in, as this identifies as a Unix PAL.

File: mcs/class/System/linux_net_4_x_System.dll.sources (hunk 1 above)
Comment (alexischr) [reply mcs/class/System/linux_net_4_x_System.dll.sources]: I'll do `corefx/Unix/Interop.Read.cs`

File: mcs/class/System/corefx/Interop.Read.cs (hunk 8 above)
Comment (alexischr) [reply mcs/class/System/corefx/Interop.Read.cs]: The point is to not have a syscall write into the managed heap

File: mcs/class/System/corefx/Interop.Read.cs (hunk 7 above)
Comment (alexischr) [reply mcs/class/System/corefx/Interop.Read.cs]: @luhenry ah, but `has_ref` will always be false if `fd.DangerousAddRef` failed :) I see that the common use of DangerousAddRef is to only use the success indicator on release, but that is not exclusive. In this case as we could be passing a recycled handle to a syscall, which cannot not be detected at any point later on as an invalid handle. So I am not convinced that I should remove this, and at worst it is harmless

File: mcs/class/System/corefx/Interop.Read.cs (hunk 2 above)
Comment (luhenry) [reply mcs/class/System/corefx/Interop.Read.cs]: No need to have the `_Internal` suffix since the signature is different anyway.

File: mcs/class/System/corefx/Interop.Read.cs [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+            bool has_ref = false;
+            try {
+                fd.DangerousAddRef (ref has_ref);
+                if (has_ref)
+                    bytes_read = Read_Internal (fd, unmanaged_buffer, count);
```
Comment (luhenry) [inline mcs/class/System/corefx/Interop.Read.cs]: Pass `fd.DangerousGetHandle()` otherwise the runtime will get an object since InternalCall do *zero* marshalling.

File: mono/metadata/pal-icalls.c (hunk 23 above)
Comment (luhenry) [reply mono/metadata/pal-icalls.c]: The runtime already tries to abort the thread, even if there is no `install`+`uninstall` around the blocking syscall, so what you are doing in `interrupt_read` is redundant with what the runtime does regardless of the presence of an interrupt token or not. Or in other words: it's not because you don't install an interrupt callback that the runtime won't send an abort signal (that's what `mono_threads_suspend_abort_syscall` does).

Comment (alexischr) [reply mono/metadata/pal-icalls.c]: > The runtime already tries to abort the thread, even if there is no install+uninstall around the bocking syscall

This is not true. I can demonstrate:

1) Build this PR on Linux, and run `mono/tests/bug-80307.exe` with that Mono
2) Comment out the `install`+`uninstall`
3) Run the test again; the test will hang.

The particular test just happens to exit without shutting down its FSW, causing this situation.

Comment (alexischr) [reply mono/metadata/pal-icalls.c]: @luhenry I see from `suspend_sync` that what you are saying maybe should happen if `interrupt_kernel` is true, however you can observe that this is not happening. 

Even if the runtime did deliver a signal automatically, this is still not enough. We need to be able to distinguish between a runtime abort interrupt, and a kernel EINTR that just wants us to do a loop. I do not believe there is a way to distinguish this in managed code.

File: mcs/class/System/corefx/Interop.Read.cs (hunk 8 above)
Comment (luhenry) [reply mcs/class/System/corefx/Interop.Read.cs]: That doesn't matter, and use the `fixed` C# keyword to make sure the buffer doesn't move in memory while passed to native.

File: mcs/class/System/corefx/Interop.Read.cs (hunk 7 above)
Comment (luhenry) [reply mcs/class/System/corefx/Interop.Read.cs]: If `fd.DangerousAddRef` fails, it throws. The only use of `has_ref` is for the finally clause, since it will execute even in case of exceptions, and that's also why you pass `has_ref` by ref so it's set even in case of exception. What you have is just not how you use SafeHandle, so just look at  https://github.com/mono/mono/blob/master/mcs/class/System/System.Net.Sockets/Socket.cs#L187 to see an example on how you use it.

File: mono/metadata/pal-icalls.c (hunk 23 above)
Comment (luhenry) [reply mono/metadata/pal-icalls.c]: You should look more closely at how and where this interrupt mechanism is used: it's used by Thread.Abort and Thread.Interrupt to be able to force the thread to exit a blocking syscall which wouldn't otherwise be interrupted by a signal. It's not here to replace the interrupting property of a signal (or the OS API we rely on), it's here to **complement** it. So sending a signal (or the OS API we rely on) **in** the interrupt callback is just double work.

Now, how does Thread.Interrupt or Thread.Abort then use this interrupt token/callback: 1. it will send the signal (or the OS API we rely on) to suspend the thread, 2. set a flag on the thread so that we throw a `ThreadAbortException` or a `ThreadInterruptedException` based on what you tried to do, and 3. use the interrupt token to call the interrupt callback (ie `interrupt_read` in this case). This "flag on the thread" is then checked in and out of internal call transitions, so that if the thread was interrupted, it can throw the proper exception.

That explanation aside, what you have currently doesn't make the difference between a EINTR which happened because of an explicit interrupt or a spurious interruption of the syscall from the point of view of managed since you are not returning "interrupted" in any case.

The gist of it is: you can safely do the `while` loop in managed without installing/uninstalling the token since it's all redundant to what the runtime is already doing.

File: mcs/class/System/corefx/Interop.Read.cs (hunk 7 above)
Comment (alexischr) [reply mcs/class/System/corefx/Interop.Read.cs]: @luhenry I missed that DangerousAddRef throws (the docs don't mention any exceptions for the method!!). You are totally right!

File: mono/metadata/pal-icalls.c (hunk 23 above)
Comment (alexischr) [reply mono/metadata/pal-icalls.c]: Thank you for the explanation, I see all of it is handled! I copied the interrupt callback from `ves_icall_System_Net_Sockets_Socket_Accept_internal`, which does use `mono_threads_suspend_abort_syscall` in the callback, but that situation must be different somehow.

> That explanation aside, what you have currently doesn't make the difference between a EINTR which happened because of an explicit interrupt or a spurious interruption of the syscall from the point of view of managed since you are not returning "interrupted" in any case.


I am using `interrupted`: 
```
} while (!interrupted && result < 0 && errno == EINTR);
```

> The gist of it is: you can safely do the while loop in managed without installing/uninstalling the token since it's all redundant to what the runtime is already doing.

I will do that :)

Comment (luhenry) [reply mono/metadata/pal-icalls.c]: Let me take a look at Accept to understand better why it's needed.

For the `interrupted`, you only use it to know if you should break the loop because you don't check if the "flag on the thread" is set. If you switch the loop to managed, the "flag on the thread" will be checked when transitioning from managed to runtime.

**Comment by alexischr** [comment]
@monojenkins build deb with monolite

**Comment by alexischr** [comment]
@monojenkins commit csproj

**Comment by alexischr** [comment]
@monojenkins commit csproj

**Comment by alexischr** [comment]
@monojenkins build deb with monolite

**Comment by alexischr** [comment]
@monojenkins commit csproj

**Comment by alexischr** [comment]
@monojenkins commit csproj

**Comment by alexischr** [comment]
@monojenkins commit csproj

**Comment by alexischr** [comment]
@luhenry could you please re-review?

File: mono/metadata/marshal.c [hunk 25]
```diff
@@ -4869,8 +4869,8 @@ mono_marshal_set_last_error (void)
 {
 	/* This icall is called just after a P/Invoke call before the P/Invoke
 	 * wrapper transitions the runtime back to running mode. */
-	MONO_REQ_GC_SAFE_MODE;
 #ifdef WIN32
+	MONO_REQ_GC_SAFE_MODE;
```
Comment (marek-safar) [inline mono/metadata/marshal.c]: Is this relevant to this patch?

Comment (lambdageek) [reply mono/metadata/marshal.c]: @marek-safar I believe without this change the PR will fail some coop/hybrid suspend tests.  But I think that if we had a Windows build with checked mode it would fail.  So it's not the best fix.

@alexischr The issue is that `mono_marshal_set_last_error` is called from at least 3 different contexts: 1. directly from filewatcher.c in GC Unsafe mode; 2. in this PR from a PAL icall from a GC Safe region; 3. as an icall in its own right in GC Unsafe mode from `marshal-ilgen.c`.

It would be better to have separate entry points for each of those uses with their own requirements that call a common utility function.

Comment (lambdageek) [reply mono/metadata/marshal.c]: Not sure if my suggestion should be part of this PR or some followup work.

Comment (alexischr) [reply mono/metadata/marshal.c]: @marek-safar @lambdageek this icall does not exist under Windows (or the situation where this is useful, which is runtime syscall abort) so I don't think it is a problem here.

This is needed because in the case the syscall is aborted by the runtime, in which case the thread state will be STATE_BLOCKING_SUSPEND_REQUESTED by the time we get to this call, which is not allowed by `assert_gc_safe_mode`. It may be that that can be added to the allowed modes?

This is just an easy fix on a check that was not necessary in the first place. It fixes the coop test failures.

**Comment by alexischr** [comment]
@monojenkins backport to 2018-06

**Comment by alexischr** [comment]
@monojenkins build deb with monolite

**Comment by alexischr** [comment]
@lewurm It did pass CI, but I saw those failures once before and could not reproduce on the PR so they are flaky or only on some hosts. 

I will investigate

**Comment by lewurm** [comment]
I don't see results where it passed. The commit that has been merged has a failing status on aarch64/linux: https://jenkins.mono-project.com/job/test-mono-pull-request-arm64/11758/
